concepts/functions/about.md

```diff
@@ -1,3 +1,302 @@
+1
+```
+
+## Modularity
+
+Complex programs can be broken down into smaller parts. Functions can be used to perform different tasks.
+
+Assume a program has to perform the following tasks:
+  * Calculate the area of a circle
+  * Calculate the area of a rectangle
+  * Calculate the area of a triangle
+
+We can break down the program into smaller parts.
+
+```python
+def circle_area(r):
+    return 3.14 * r * r
+
+def rectangle_area(length, breadth):
+    return length * breadth
+
+def triangle_area(base, height):
+    return 0.5 * base * height
+```
+
+Now, we can call the functions in the order we want.
+
+```python
+>>> circle_area(2)
+12.56
+>>> rectangle_area(2, 3)
+6
+>>> triangle_area(2, 3)
+1.5
+```
+
+## Scope of Variables
+
+If variable is defined inside a function, then it will be only accessible inside the function. If we want to access the variable outside the function, we need to use the [`global`][global] keyword. [`nonlocal`][nonlocal] keyword is used to access the variable inside a nested function.
+
+```python
+def outer():
+    x = 10
+    print('Inside function:', x)
+```
+
+variable `x` is defined inside the outer function. It is limited to the scope of the outer function only. It will not alter the value of the variable outside the outer function.
+
+```python
+>>> x = 30
+
+# regardless of whether we call the function or not, the value of x will be 30
+>>> outer()
+Inside function: 10
+>>> x
+30
+```
+
+We can access the variable inside the outer function using the `global` keyword.
+```python
+def outer():
+    global x
+    x = 10
+    print('Inside function:', x)
+```
+
+As we have used the `global` keyword, the value of `x` will be changed.
+
+```python
+>>> x = 30
+>>> outer()
+Inside function: 10
+>>> x
+10
+```
+
+
+## Functions as first class objects
+
+In python, functions can be assigned to variables and passed as arguments to other functions. They can also be used as return values in other functions. Functions can also be used as items in a sequence([`list`][list], [`tuple`][tuple] etc) or as value in [`dict`][dict]. This is called _function as [`first class object`][first class objects]_.
+
+
+```python
+# print is a function
+
+# function can be assigned to a variable
+>>> fake_print = print
+>>> fake_print("Hello")
+Hello
+>>> type(fake_print)
+<class 'builtin_function_or_method'>
+
+# function can be passed as an argument to another function
+>>> def check_print(func):
+    func("Hello")
+>>> check_print(print)
+Hello
+
+# function can be used as an item in a sequence
+>>> my_list = [print, "Hello"]
+>>> my_list[0]("Hello")
+Hello
+
+# function can be used as a value in a dictionary
+>>> my_dict = {"key": print}
+>>> my_dict["key"]("Hello")
+Hello
+
+# function can be used as a return value in another function
+>>> def return_func():
+    return print
+>>> return_func()("Hello")
+Hello
+```
+
+Functions can also be nested inside other functions.
+```python
+def outer():
+    x = 10
+    def inner():
```

Comment:
This is not used often and is often not good practice. Maybe it should be skipped.

Reply:
Please read the connected issue before reviewing. `nested functions` is in the concepts of #2353. Keep nested functions in the exercise.

Nested functions are not bad practice at all. They can be used to access the enclosing namespace. If such function is only used by one other function, then you could opt in to use `nested` or `inner functions`.

Reply:
If you really would like these gone. Throw up an issue on the repo. We could either then put this issue on hold, or merge this issue (if everything else is okay) and then change the docs afterwards.

Reply:
I think [issue 2353](https://github.com/exercism/python/issues/2353) is pretty clear. The intent of this exercise is to dig deeper into Python functions - both as a "next step" for students and as an introduction into `higher order functions` and FP.  As well as more practice before students are "hit" with all the function parameter options, and unpacking actions (function arguments exercise).

Python's object model (as opposed to Javas object model, for example) treats `function objects` as first-class citizens.  It is possible (perhaps not _advisable_) to write nested functions -- or even constructs  where one function encloses multiple other functions. Students need to know that -- even if they don't end up using it all that often.

And that also helps teach scope (along with locals and globals) -- which (we hope) leads eventually to understanding what a `closure` is, and why you'd want one, and other things like `decorators`.

Reply:
This was not intended as a "basic" exercise.  We have `lasgana` -- and all the exercises in-between for that.  Students would not be taking on this exercise until after they finished all the exercises on data structures and sequences and more.